### PR TITLE
Create Retrieval Augmented Generation - Contract Analyst.ipynb

### PR DESCRIPTION
**Pull request by guruvittal**
Using RAG for contract analysis

### CODE REVIEW

File: language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb [hunk 1]
```diff
@@ -0,0 +1,717 @@
+{
```
Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <p>It doesn't look like this Notebook uses the Generative AI for Developers PaLM API or MakerSuite. This repository primarily focuses on Google Cloud, so I'd recommend removing this.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #151.            # print(f'Sending embedding request for: {head!r}')
</pre><p>Remove this commented out line</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <p>Remove the actual Project ID and replace with a placeholder</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.    from langchain.vectorstores import Chroma
</pre><p>Put all imports at the top of the notebook as well</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <p>Can you add some more descriptions/explanations for each of these headers?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <p>Remove some of these unnecessary output sections </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #151.            # print(f'Sending embedding request for: {head!r}')
</pre><p>Remove this commented out line</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    !pip install -q chromadb
</pre><p>Put all of these installs at the top of the notebook</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.    from langchain.vectorstores import Chroma
</pre><p>Put all imports at the top of the notebook as well</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <p>Can you add some more descriptions/explanations for each of these headers?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <p>Remove some of these unnecessary output sections </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

**Comment by holtskinner** [comment]
<p>General comment, run black formatting on this notebook. Install nox <a href="https://pypi.org/project/nox" target="_blank">https://pypi.org/project/nox/</a> and run this from the top directory of the repository. (Don't include other files that aren't relevant in your PR)</p><p> </p><pre class="ql-syntax" spellcheck="false">nox -s blacken
</pre>

File: language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb (hunk 1 above)
Comment (gprakash) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

File: language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
```
Comment (holtskinner) [inline language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <p>Could you add a short description about what this front end does?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/210/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

[the remark this answers is not in the document]
Comment (gprakash) [reply language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: <p> </p><p>On the llm library from LangChain:</p><ul><li>The VertexAI() call requires a stop sequence.</li><li>I don't need a stop sequence so we will have to override the generate method to make stop sequence optional.</li><li>Hence I am using VertexLLM() which is a simple tweak of the base model.</li></ul><p> </p><p>On the embedding library from LangChain:</p><ul><li>The langchain library would not support rate limiting - hence I included a method with rate limiting - I ran into rate limits few weeks back. But I am not facing those issues now - so I can use the standard langchain embedding class. I can update the patch with this change and re-submit.</li></ul><p> </p>

[the remark this answers is not in the document]
Comment (holtskinner) [reply language/orchestration/langchain/Retrieval Augmented Generation - Contract Analyst.ipynb:1]: Thanks for the explanations! For the `VertexAI()` class, maybe you could make a Pull Request to LangChain to make the stop sequence optional as well? I won't block this Notebook based on that, but could be a good change to make to the library.

**Comment by holtskinner** [comment]
I ran the formatter for you and I moved the file into the retrieval augmented generation folder for organization.

**Review by holtskinner — APPROVED** [review]

**Pull request merged**